src/Parallel/LockfreeUnorderedSet.hpp

```diff
@@ -0,0 +1,451 @@
+  // NOLINTNEXTLINE(modernize-avoid-c-arrays)
+  std::byte padding_[cache_line_size_ - 2 * sizeof(std::uint64_t) -
+                     sizeof(decltype(entries_))] = {};
+};
+
+template <class T, std::uint64_t EmptySlotValue, bool TrackBucketSize>
+FixedSizeUnorderedSet<T, EmptySlotValue, TrackBucketSize>::
+    FixedSizeUnorderedSet(const size_t number_of_buckets)
+    : entries_(new(std::align_val_t{cache_line_size_})
+                   std::atomic<std::uint64_t>[number_of_buckets * bucket_size]),
+      number_of_buckets_(number_of_buckets) {
+  if (number_of_buckets_ == 0) {
+    ERROR("The capacity must be a power of two larger than 0. Got "
+          << number_of_buckets_);
+  }
+  if ((number_of_buckets_ bitand (number_of_buckets_ - 1)) != 0) {
+    ERROR("The capacity must be a power of two larger than 0. Got "
+          << number_of_buckets_);
+  }
+  const auto uint64_log2 = [](uint64_t n) -> std::uint64_t {
+    int i = -static_cast<int>(n == 0);
+
+    const auto helper = [&i, &n](auto k) -> void {
+      if (n >= (static_cast<std::uint64_t>(1) << decltype(k)::value)) {
+        i += static_cast<int>(decltype(k)::value);
+        n >>= decltype(k)::value;
+      }
+    };
+    helper(std::integral_constant<size_t, 32>{});
+    helper(std::integral_constant<size_t, 16>{});
+    helper(std::integral_constant<size_t, 8>{});
+    helper(std::integral_constant<size_t, 4>{});
+    helper(std::integral_constant<size_t, 2>{});
+    helper(std::integral_constant<size_t, 1>{});
+    return static_cast<std::uint64_t>(i);
+  };
+  dimension_ = uint64_log2(number_of_buckets_);
+
+  // Explicitly zero out the counters.
+  for (size_t i = 0; i < number_of_buckets_ * bucket_size; ++i) {
+    entries_[i].store(EmptySlotValue, std::memory_order_relaxed);
+  }
+}
+
+template <class T, std::uint64_t EmptySlotValue, bool TrackBucketSize>
+bool FixedSizeUnorderedSet<T, EmptySlotValue, TrackBucketSize>::insert(
+    const T key) noexcept {
+  const std::uint64_t internal_key = compute_internal_key(key);
+  const std::uint64_t bucket_index = hash(internal_key) * bucket_size;
+  if constexpr (TrackBucketSize) {
+    std::atomic<std::uint64_t>& this_bucket_size = entries_[bucket_index];
+    // We first do a linear search in the bucket to see if the element exists.
+    for (size_t i = 0; i < this_bucket_size.load(std::memory_order_relaxed);
+         ++i) {
+      const size_t index = bucket_index + 1 + i;
+      const std::uint64_t probed_internal_key =
+          entries_[index].load(std::memory_order_relaxed);
+      if (probed_internal_key == internal_key) {
+        return true;  // set already contains value
+      }
+    }
+
+    // We now know that _if_ the element exists, it's in what was in the first
+    // empty slot, which means we either encounter the element in the first
+    // available slot or we get to insert it in that slot.
+    for (size_t i = this_bucket_size.load(std::memory_order_relaxed);
```

Comment:
Since this loop starts at the bucket size and that is never decreased, it looks like this implementation can never reuse slots.  I haven't tested, but I think that means just adding and removing the same value 8 times will cause the last insertion to fail.